src/main/scala/cognite/spark/v1/FlexibleDataModelRelationUtils.scala

```diff
@@ -668,58 +665,81 @@ object FlexibleDataModelRelationUtils {
       case Success(i) => get(i).map(FieldSpecified(_))
     }
 
+  private def getListPropAsSeq[T](propertyName: String, row: Row, index: Integer): Seq[T] =
+    Try(row.getSeq[T](index)) match {
+      case Success(x) => x
+      case Failure(_) =>
+        Try(row.getAs[Array[T]](index).toSeq) match {
+          case Success(x) => x
+          case Failure(err) =>
+            throw new CdfSparkException(
+              f"""Could not deserialize property $propertyName as an array of the expected type""",
+              err)
+        }
+    }
+
   // scalastyle:off cyclomatic.complexity method.length
   private def toInstancePropertyValueOfList(
       row: Row,
       schema: StructType,
       propertyName: String,
-      propDef: CorePropertyDefinition): Either[Throwable, OptionalField[InstancePropertyValue]] =
+      propDef: CorePropertyDefinition,
+      instanceSpace: Option[String]
+  ): Either[Throwable, OptionalField[InstancePropertyValue]] =
     lookupFieldInRow(row, schema, propertyName, propDef.nullable.getOrElse(true)) { i =>
       propDef.`type` match {
-        case p: TextProperty if p.isList =>
-          val strSeq = Try(row.getSeq[Any](i)).getOrElse(row.getAs[Array[Any]](i).toSeq)
-          Try(InstancePropertyValue.StringList(skipNulls(strSeq).map(String.valueOf))).toEither
-        case p @ PrimitiveProperty(PrimitivePropType.Boolean, _) if p.isList =>
-          val boolSeq = Try(row.getSeq[Boolean](i)).getOrElse(row.getAs[Array[Boolean]](i).toSeq)
-          Try(InstancePropertyValue.BooleanList(boolSeq)).toEither
-        case p @ PrimitiveProperty(PrimitivePropType.Float32, _) if p.isList =>
-          val floatSeq = Try(row.getSeq[Any](i)).getOrElse(row.getAs[Array[Any]](i).toSeq)
-          tryAsFloatSeq(floatSeq, propertyName)
+        case _: DirectNodeRelationProperty =>
+          val structSeq = getListPropAsSeq[Row](propertyName, row, i)
+          tryAsDirectNodeRelationList(structSeq, propertyName, instanceSpace)
+            .map(InstancePropertyValue.ViewDirectNodeRelationList)
+        case _: TextProperty =>
+          Try({
+            val strSeq = getListPropAsSeq[String](propertyName, row, i)
+            InstancePropertyValue.StringList(skipNulls(strSeq))
```

Comment:
for TextProperty it was looking a bit differently before - used `[Any]` and `.map(String.valueOf)`. In case of String class it should behave the same as before but not sure if we could have TextProperty with some other type inside :thinking:

Reply:
I think it's ok because it's wrapped in the try. Maybe it would make more sense to modify the other ones to use a similar construct too?

Reply:
In case there can be other type inside, those cases would now generate errors. Most likely it is always string but keeping Any to be on a safe side is an option

Reply:
Ah though not actually, the behavior will differ, since String.valueOf cannot fail